robot/impl/local_robot_test.go

```diff
@@ -4239,3 +4240,75 @@ func TestMaintenanceConfig(t *testing.T) {
 		test.That(t, sensorUnBlocked, test.ShouldNotBeNil)
 	})
 }
+
+func TestRemovingOfflineRemote(t *testing.T) {
+	logger, _ := logging.NewObservedTestLogger(t)
+	ctx := context.Background()
+
+	motorName := "remoteMotorFoo"
+	motorResourceName := resource.NewName(motor.API, motorName)
+
+	remoteCfg := &config.Config{
+		Components: []resource.Config{
+			{
+				Name:                motorName,
+				Model:               resource.DefaultModelFamily.WithModel("fake"),
+				API:                 motor.API,
+				ConvertedAttributes: &fakemotor.Config{},
+			},
+		},
+	}
+
+	remoteRobot := setupLocalRobot(t, ctx, remoteCfg, logger.Sublogger("remote"))
+	remoteOptions, _, remoteAddr := robottestutils.CreateBaseOptionsAndListener(t)
+	err := remoteRobot.StartWeb(ctx, remoteOptions)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Set up a local main robot which is connected to the remote.
+	mainRobotCfg := &config.Config{
+		Remotes: []config.Remote{
+			{
+				Name:    "remote",
+				Address: remoteAddr,
+				// These values dictate how quickly we'll observe the remote going offline. And how
+				// quickly we'll observe it coming back online.
+				ConnectionCheckInterval: 10 * time.Millisecond,
+				ReconnectInterval:       10 * time.Millisecond,
+			},
+		},
+	}
+	mainRobotI := setupLocalRobot(t, ctx, mainRobotCfg, logger.Sublogger("main"))
+	// We'll manually access the resource manager to move the test forward.
+	mainRobot := mainRobotI.(*localRobot)
+
+	// Grab the RobotClient resource graph node from the main robot that is connected to the
+	// remote. We'll use this to know when the main robot observes the remote has gone offline.
+	mainToRemoteClientRes, _ := mainRobot.RemoteByName("remote")
+	test.That(t, mainToRemoteClientRes, test.ShouldNotBeNil)
+	mainToRemoteClient := mainToRemoteClientRes.(*client.RobotClient)
+	test.That(t, mainToRemoteClient.Connected(), test.ShouldBeTrue)
+
+	// Stop the remote's web server. Wait for the main robot to observe there's a connection problem.
+	logger.Info("Stopping web")
+	remoteRobot.StopWeb()
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		tb.Helper()
+		test.That(tb, mainToRemoteClient.Connected(), test.ShouldBeFalse)
+	})
+	// Set node status to unhealthy
+	remotes := mainRobot.manager.resources.FindNodesByAPI(client.RemoteAPI)
+	node, _ := mainRobot.manager.resources.Node(remotes[0])
+	node.LogAndSetLastError(errors.New("Set Node to unhealthy"))
+	
+	// Reconfigure the main robot with the offline remote removed
+	mainRobot.Reconfigure(ctx, &config.Config{})
+
+	// Ensure that the remote has been remoted correctly
+	findRemote, ok := mainRobot.RemoteByName("remote")
+	test.That(t, findRemote, test.ShouldBeEmpty)
+	test.That(t, ok, test.ShouldBeFalse)
+
+	// Ensure that motor is removed
+	names := mainRobot.ResourceNames()
+	test.That(t, names, test.ShouldNotContain, motorResourceName)
```

Comment:
I am trying to recreate the situation where the main robot is trying to connect to the unhealthy remote while also reconfiguring to remove the remote. 

What I have is close to to doing that but it I tried it with and without the fix in place and it seems to pass in either case

Reply:
@maximpertsov @cheukt

Reply:
though looking at what we're trying to do, setting the node as unhealthy first may not be the right way to go about it - do remotes get set as unhealthy when it goes offline?

Reply:
the remote become unhealthy through connection error after they are offline: https://github.com/viamrobotics/rdk/blob/main/robot/impl/resource_manager.go#L812

Reply:
My hope was we could short cut the unhealthy status so we dont have to wait on the connection issue. That way we can test removing an unhealthy remote

Reply:
That said It looks like setting it doesnt really change anything so I think ill just remove the setting unhealthy.